crates/phactory/src/system/gk.rs

```diff
@@ -261,50 +252,47 @@ where
         }
     }
 
-    fn process_contract_event(
+    fn process_cluster_event(
         &mut self,
         origin: MessageOrigin,
-        event: ContractEvent<chain::Hash, chain::AccountId>,
+        event: ClusterEvent,
     ) -> Result<(), TransactionError> {
-        info!("Incoming contract event: {:?}", event);
+        info!("Incoming cluster event: {:?}", event);
         match event {
-            ContractEvent::InstantiateCode {
-                contract_info,
-                deploy_workers,
-            } => {
+            ClusterEvent::DeployCluster { cluster, workers } => {
                 if !origin.is_pallet() {
-                    error!("Attempt to instantiate pink from bad origin");
+                    error!("Attempt to deploy cluster from bad origin");
                     return Err(TransactionError::BadOrigin);
                 }
 
-                // first, update the on-chain ContractPubkey
-                let contract_id = contract_info.contract_id(Box::new(hashing::blake2_256));
-                let contract_key = get_contract_key(&self.master_key, &contract_info);
+                // first, update the on-chain cluster pubkey
+                let cluster_key = get_cluster_key(&self.master_key, &cluster);
+                let cluster_pubkey = cluster_key.public();
                 self.egress
-                    .push_message(&ContractRegistryEvent::PubkeyAvailable {
-                        contract: contract_id,
-                        pubkey: contract_key.public(),
+                    .push_message(&ClusterRegistryEvent::PubkeyAvailable {
+                        cluster,
+                        pubkey: cluster_pubkey,
                     });
-                // then distribute contract key to each worker
-                // and update the on-chain deployment state
+                // then distribute cluster key to each worker
+                // the on-chain deployment state should be updated by assigned workers
                 let ecdh_key = self
                     .master_key
                     .derive_ecdh_key()
-                    .expect("should never fail with valid master key; qed.");
+                    .expect("should never fail with valid key; qed.");
                 let secret_mq = SecretMessageChannel::new(&ecdh_key, &self.egress);
                 // TODO.shelven: set up expiration
-                for worker in deploy_workers.iter() {
+                for worker in workers.iter() {
                     secret_mq
                         .bind_remote_key(Some(&worker.ecdh_pubkey.0))
-                        .push_message(&ContractKeyDistribution::contract_key_distribution(
-                            contract_key.dump_secret_key(),
-                            contract_info.clone(),
+                        .push_message(&ClusterKeyDistribution::cluster_key_distribution(
```

Comment:
Messages might arrive the chain at different block. That might cause the clusters in different workers diverge?
If we deploy contract subsequently, the cluster in some workers are ready, some are not.

Reply:
Uploading code is now only allowed when cluster is deployed to all the workers. I do not make the same check in instantiation since it will certainly fail before the code uploading.

Reply:
> Uploading code is now only allowed when cluster is deployed to all the workers.

This solution doesn't feel perfect. It introduce kinda implict requirement for the system to work properly. It is easy to ignore the requirement in the future when someone design a new feature.

Reply:
> Uploading code is now only allowed when cluster is deployed to all the workers.

This solution doesn't feel perfect. It introduce kinda implict requirement for the system to work properly. It is easy to ignore the requirement in the future when someone design a new feature or refactor something.

Reply:
Implemented `BatchDispatchClusterKeyEvent`